lib/nat/service.go

```diff
@@ -372,3 +405,27 @@ func hash(input string) int64 {
 	h.Write([]byte(input))
 	return int64(h.Sum64())
 }
+
+func addrSetsEqual(a []Address, b []Address) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	// TODO: Rewrite this using slice.Contains once Go 1.21 is the minimum Go version.
+	for _, aElem := range a {
+		aElemFound := false
+		for _, bElem := range b {
+			if bElem.Equal(aElem) {
+				aElemFound = true
+				break
+			}
+		}
+		if !aElemFound {
+			// Found element in a that is not in b.
+			return false
+		}
+	}
```

Comment:
This is quadratic for unequal sets which is not great, even though I expect them to be small in practice. I'd either sort the sets (somewhere) if the order isn't critical and then `slices.Equal`, or use a map for lookups.

Reply:
I'm aware of the theoretical time complexity of this algorithm as `len(a)` or `len(b)` tend to infinity but honestly I don't think it matters too much here. Sorting or using a map would require us to define an order on IP addresses which seems unclean and overkill to me.

Given that the set size is at most by the number of addresses on an interface and that this code is not in the "hot path" of anything remotely performance-critical, I don't really expect this to cause any sort of slowness for any realistic setup. Even if you had, say, 100 IP addresses on an interface, this loop would only execute at most 10.000 times, which takes virtually no time on a modern machine, especially compared to the network calls involved in doing this so I'd personally try to avoid premature optimization here.